crates/cli/cli/src/commands/query/connection/end.rs

```diff
@@ -0,0 +1,91 @@
+use hermes_cli_framework::output::Output;
+use oneline_eyre::eyre::eyre;
+use tracing::error;
+use tracing::info;
+
+use hermes_cli_framework::command::Runnable;
+use hermes_cosmos_client_components::traits::chain_handle::HasBlockingChainHandle;
+use hermes_cosmos_relayer::contexts::builder::CosmosBuilder;
+use hermes_cosmos_relayer::types::error::BaseError as RelayerError;
+use ibc_relayer::chain::handle::ChainHandle;
+use ibc_relayer::chain::requests::QueryConnectionRequest;
+use ibc_relayer::chain::requests::{IncludeProof, QueryHeight};
+use ibc_relayer_types::core::ics03_connection::connection::State;
+use ibc_relayer_types::core::ics24_host::identifier::ChainId;
+use ibc_relayer_types::core::ics24_host::identifier::ConnectionId;
+use ibc_relayer_types::Height;
+
+use crate::Result;
+
+#[derive(Debug, clap::Parser)]
+pub struct QueryConnectionEnd {
+    #[clap(
+        long = "chain",
+        required = true,
+        value_name = "CHAIN_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the chain to query"
+    )]
+    chain_id: ChainId,
+
+    #[clap(
+        long = "connection",
+        visible_alias = "conn",
+        required = true,
+        value_name = "CONNECTION_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the connection to query"
+    )]
+    connection_id: ConnectionId,
+
+    #[clap(
+        long = "height",
+        value_name = "HEIGHT",
+        help = "Height of the state to query. Leave unspecified for latest height."
+    )]
+    height: Option<u64>,
+}
+
+impl Runnable for QueryConnectionEnd {
+    async fn run(&self, builder: CosmosBuilder) -> Result<Output> {
+        let chain = builder.build_chain(&self.chain_id).await?;
+        let chain_id = self.chain_id.clone();
+        let connection_id = self.connection_id.clone();
+        let height = self.height;
+
+        let connection_end = chain
+            .with_blocking_chain_handle(move |chain_handle| {
```

Comment:
Minor: If possible, let's try to do the implementation abstractly so that it can be reused or refactored later on.

What we would need is to first define an abstract `ConnectionEnd` type:

```rust
#[derive_component(ConnectionEndTypeComponent, ProvideConnectionEndType)]
pub trait HasConnectionEndType<Counterparty>: Async {
  type ConnectionEnd: Async;
}
```

And a query trait:

```rust
#[derive_component(ConnectionEndQuerierComponent, ConnectionEndQuerier<Chain>)]`
pub trait CanQueryConnectionEnd<Counterparty>: 
  HasConnectionEndType<Counterparty> + HasIbcChainTypes<Counterparty> + HasErrorType
{
  async fn query_connection_end(&self, connection_id: &Self::ConnectionId, height: Option<&Self::Height>) -> Result<Self::ConnectionEnd, Self::Error>;
}
```

We would then add a context-generic Cosmos implementation in `cosmos-client-components` as follows:

```rust
pub struct QueryConnectionEndWithChainHandle;

impl<Chain, Counterparty> ConnectionEndQuerier<Chain, Counterparty>
  for QueryConnectionEndWithChainHandle
where
  Chain: HasConnectionEndType<ConnectionEnd = ConnectionEnd> 
    + HasIbcChainTypes<Counterparty, Height = Height, ConnectionId = ConnectionId>
    + HasBlockingChainHandle
{
  async fn query_connection_end(
    chain: &Chain,
    connection_id: &ConnectionId,
    height: Option<&Height>
  } -> Result<ConnectionEnd, Chain::Error> {
    ...
  }
}
```

The rest is mainly to add the wiring for the implementations to component graphs such as `CosmosClientComponents` and `CosmosChainComponents`.

This is not urgent and can be done in separate PR. But it could serve as a good exercise to get familiar with the component system in Hermes SDK.

Reply:
This would be an initial step to abstracting the CLI so that it can be more easily implemented for non-Cosmos chains?

Reply:
I opened an issue to follow-up on this https://github.com/informalsystems/hermes-sdk/issues/153

Reply:
> This would be an initial step to abstracting the CLI so that it can be more easily implemented for non-Cosmos chains?

Yup. Let's follow up on #153 then.